File: mbfpp/code/ScalaExamples/src/main/scala/com/mblinn/mbfpp/oo/fi/PersonExpanded.scala
package com.mblinn.mbfpp.oo.fi.personexpanded

object PersonExample {
  case class Person(firstName: String, middleName: String, lastName: String)
  val p1 = Person("Aaron", "Jeffrey", "Smith")
  val p2 = Person("Aaron", "Bailey", "Zanthar")
  val p3 = Person("Brian", "Adams", "Smith")
  val people = Vector(p1, p2, p3)

  def complicatedSort(p1: Person, p2: Person) =
    if (p1.firstName != p2.firstName) 
      p1.firstName < p2.firstName
    else if (p1.lastName != p2.lastName)
      p1.lastName < p2.lastName
    else
      p1.middleName < p2.middleName

}

File: mbfpp/code/ClojureExamples/src/mbfpp/rso/person_expanded.clj
(ns mbfpp.rso.person-expanded)

(def p1 {:first-name "Aaron" :middle-name "Jeffrey" :last-name "Smith"})
(def p2 {:first-name "Aaron" :middle-name "Bailey" :last-name "Zanthar"})
(def p3 {:first-name "Brian" :middle-name "Adams" :last-name "Smith"})
(def people [p1 p2 p3])
(defn complicated-sort [p1 p2]
  (let [first-name-compare (compare (p1 :first-name) (p2 :first-name))
        middle-name-compare (compare (p1 :middle-name) (p2 :middle-name))
        last-name-compare (compare (p1 :last-name) (p2 :last-name))]
    (cond 
      (not (= 0 first-name-compare)) first-name-compare
      (not (= 0 last-name-compare)) last-name-compare
      :else middle-name-compare)))
